### PR TITLE
[K/N] Prohibit empty ObjCNames

### PR DESCRIPTION
**Pull request by rickclephas**
Empty names aren't handled correctly in the ObjC export and should be prohibited (for now).

cc @SvyatoslavScherbina

### CODE REVIEW

File: kotlin-native/backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/objcexport/ObjCExportNamer.kt [hunk 1]
```diff
@@ -1016,8 +1015,8 @@ private fun KtClassOrObject.getObjCName(): ObjCName {
         for (argument in annotation.valueArguments) {
             val argName = argument.getArgumentName()?.asName?.asString() ?: (argNames - processedArgs).firstOrNull() ?: break
             when (argName) {
-                "name" -> objCName = argument.getStringValue()
-                "swiftName" -> swiftName = argument.getStringValue()
+                "name" -> objCName = argument.getStringValue()?.takeIf { it.isNotBlank() }
+                "swiftName" -> swiftName = argument.getStringValue()?.takeIf { it.isNotBlank() }
```
Comment (SvyatoslavScherbina) [inline kotlin-native/backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/objcexport/ObjCExportNamer.kt:1019]: > Empty or blank names aren't allowed, they are meant to be interpreted as `null`.

Why?
Aren't blank names valid for Objective-C selector fragments? E.g.
```objective-c
- (void)bar:(int)x :(int)y;
```

Comment (rickclephas) [reply kotlin-native/backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/objcexport/ObjCExportNamer.kt:1019]: Yeah I believe that's allowed. Anyway it wasn't the intention to support this :).
Mainly because:
- there is no way to specify a `null` value for the name
- the biggest use case for ObjC, removing the duplicate parameter name ([KT-48076](https://youtrack.jetbrains.com/issue/KT-48076)), is supported by explicitly specifying the function name
- it's not supported in Kotlin and Swift (which uses `_`)

We could consider supporting empty names, but that would require some other changes / checks.

Comment (SvyatoslavScherbina) [reply kotlin-native/backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/objcexport/ObjCExportNamer.kt:1019]: Ok. Although,
> there is no way to specify a `null` value for the name

Doesn't sound like a problem: one can simply omit the corresponding argument of `@ObjCName` to achieve the same.

> We could consider supporting empty names, but that would require some other changes / checks.

Can we prohibit using blank strings in `@ObjCName`, instead of treating them as unspecified?

Comment (rickclephas) [reply kotlin-native/backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/objcexport/ObjCExportNamer.kt:1019]: > one can simply omit the corresponding argument of @ObjCName to achieve the same.

Omitting the arguments means it will use the default value, which is an empty string.
It feels weird to have a different behaviour based on if the argument was specified explicitly or not.
Also in KSP the only way to identify if an argument is specified, is by comparing it to the default value.
See https://github.com/google/ksp/issues/674#issuecomment-1098549404. Which is talking about JVM since on native there is an issue where default arguments aren't even considered: https://github.com/google/ksp/issues/885.

> Can we prohibit using blank strings in @ObjCName, instead of treating them as unspecified?

If we check on empty strings instead of blank strings (which is probably better anyway), then blank strings will be prohibited by the checker since a space isn't a valid character.

Comment (SvyatoslavScherbina) [reply kotlin-native/backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/objcexport/ObjCExportNamer.kt:1019]: > It feels weird to have a different behaviour based on if the argument was specified explicitly or not.

Yes, but it is still not as bad as if we had both allowed, and that would lead to different names.

> Also in KSP the only way to identify if an argument is specified, is by comparing it to the default value.

Ok. How does that prevent us from prohibiting `""` in `ObjCName`?

> If we check on empty strings instead of blank strings (which is probably better anyway), then blank strings will be prohibited by the checker since a space isn't a valid character.

Yes, I meant empty strings of course, sorry for the confusion.


Motivation for prohibiting empty strings: this will allow us to change the behaviour later (if we decide so) without breaking changes. Prohibiting doesn't seem to break any meaningful use case so far.

Comment (rickclephas) [reply kotlin-native/backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/objcexport/ObjCExportNamer.kt]: Got it. Updated the checkers to prohibit empty names.
Added some more tests to verify that both empty and blank names are prohibited and that not specifying one of the names is still allowed.

**Review by SvyatoslavScherbina — APPROVED** [review]
Thank you!
